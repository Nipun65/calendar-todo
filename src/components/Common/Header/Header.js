import { useEffect, useState } from 'react';
import styles from './Header.module.css';
import Dropdown from '../Dropdown/Dropdown';
import Button from '../../UI/Button/Button';

function Header({
  date,
  setSelectedOption,
  selectedOption,
  view,
  setView,
  setValue,
  setToday,
}) {
  const [year, setYear] = useState(date.year);
  const [month, setMonth] = useState(date.month);
  let yearDuplicate = year;
  let monthDuplicate = month;
  useEffect(() => {
    setMonth(date.month);
    setYear(date.year);
  }, [date.month, date.year]);
  const Month = [
    'January',
    'February',
    'March',
    'April',
    'May',
    'June',
    'July',
    'August',
    'September',
    'October',
    'November',
    'December',
  ];
  const dropdownlList = ['Month', 'Year'];

  const dateObj = {
    year,
    month,
    date: date.date,
    currentYear: date.currentYear,
    currentMonth: date.currentMonth,
    selectedDate: date.selectedDate,
  };

  const selectedOptionHandler = (value) => {
    setSelectedOption(value);
  };

  // handling the back year conditionally if it's in date view then update the date object here or in other view update values in parent component
  const backYearHandler = () => {
    if (view === 'months' || view === 'years' || view === 'multiyears') {
      setValue('back');
    }
    yearDuplicate = year;
    setYear(--yearDuplicate);
    dateObj.year = yearDuplicate;
    setValue({ ...dateObj });
  };

  // handling the forward year conditionally if it's in date view then update the date object here or in other view update values in parent component
  const forwardYearHandler = () => {
    if (view === 'months' || view === 'years' || view === 'multiyears') {
      setValue('forward');
    }
    yearDuplicate = year;
    setYear(++yearDuplicate);
    dateObj.year = yearDuplicate;
    setValue({ ...dateObj });
  };

  // handling the back month conditionally if it's in date view then update the date object here or in other view update values in parent component
  const backMonthHandler = () => {
    if (view === 'months' || view === 'years' || view === 'multiyears') {
      setValue('backmonth');
    } else {
      if (month === 0) {
        setMonth(11);
        yearDuplicate = year;
        setYear(--yearDuplicate);
        dateObj.month = 11;
        dateObj.year = yearDuplicate;
      } else {
        monthDuplicate = month;
        setMonth(--monthDuplicate);
        dateObj.month = monthDuplicate;
      }
      setValue({ ...dateObj });
    }
  };

  // handling the forward month conditionally if it's in date view then update the date object here or in other view update values in parent component
  const forwardMonthHandler = () => {
    if (view === 'months' || view === 'years' || view === 'multiyears') {
      setValue('forwardmonth');
    } else {
      if (month === 11) {
        setMonth(0);
        yearDuplicate = year;
        setYear(++yearDuplicate);
        dateObj.month = 0;
        dateObj.year = yearDuplicate;
      } else {
        monthDuplicate = month;
        setMonth(++monthDuplicate);
        dateObj.month = monthDuplicate;
      }
      setValue({ ...dateObj });
    }
  };

  // setting the view
  const viewHandler = () => {
    if (view === 'date') {
      setView('months');
    } else if (view === 'months') {
      setView('years');
    } else if (view === 'years') {
      setView('multiyears');
    }
  };

  // to get today's date
  const handleToday = () => {
    setToday();
  };

  const getYearData = () => {
    let textContent = '';
    if (view === 'date' && selectedOption !== 'Year') {
      textContent += `${Month[month]} `;
    }
    if (view === 'years') {
      if (year % 10 !== 0) {
        textContent += `${year - (year % 10) + 1} - ${
          year - (year % 10) + 10
        } `;
      } else {
        textContent += `${year - 9}-${year} `;
      }
    }
    if (view === 'date' || view === 'months') {
      textContent += `${year} `;
    }
    if (view === 'multiyears') {
      textContent += `${year - (year % 100) + 1} - ${
        year - (year % 100) + 100
      } `;
    }
    return textContent;
  };

  return (
    <div className={styles.header}>
      <div className={`${styles.btnHover} ${styles['first-div']}`}>
        {view !== 'multiyears' && (
          <Button
            onClick={backYearHandler}
            classes={styles.btn}
            textContent="<<"
          />
        )}
        {selectedOption !== 'Year' && (
          <Button
            onClick={backMonthHandler}
            classes={styles.btn}
            disabled={date.year === 1}
            textContent="<"
          />
        )}
        <div className={styles['todaybtn-div']}>
          <Button
            onClick={handleToday}
            classes={styles.todaybtn}
            textContent="Today"
          />
        </div>
      </div>
      <div
        className={`
         ${
           view !== 'multiyears' && selectedOption !== 'Year'
             ? styles.btnHover
             : ''
         } ${styles['second-div']}`}
      >
        <Button
          onClick={viewHandler}
          classes={styles.btn}
          textContent={getYearData()}
          disabled={view === 'multiyears' || selectedOption === 'Year'}
        />
      </div>

      <div className={`${styles['third-div']} ${styles.btnHover}`}>
        <div className={styles.dropdown}>
          <Dropdown
            view={view}
            option={dropdownlList}
            setValue={selectedOptionHandler}
          />
        </div>

        {selectedOption !== 'Year' && (
          <Button
            onClick={forwardMonthHandler}
            classes={styles.btn}
            textContent=">"
          />
        )}
        {view !== 'multiyears' && (
          <Button
            onClick={forwardYearHandler}
            classes={styles.btn}
            textContent=">>"
          />
        )}
      </div>
    </div>
  );
}
export default Header;
