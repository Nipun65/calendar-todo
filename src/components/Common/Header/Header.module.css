.header {
  background-color: aqua;
  display: flex;
  justify-content: space-between;
  align-items: center;
  height: 70px;
}

.btn {
  background-color: transparent;
  border: none;
  font-size: 23px;
}

.btnHover :hover {
  background-color: #72a3a6;
}

ul {
  background-color: white !important;
  transition: none !important;
}

.dropdown {
  float: left;
}
.first-div {
  order: 1;
  flex-basis: 25%;
  text-align: left;
  display: inline-flex;
  justify-content: space-evenly;
}
.second-div {
  order: 2;
  flex-basis: 50%;
  text-align: center;
}
.third-div {
  order: 3;
  flex-basis: 25%;
  text-align: right;
  display: inline-flex;
  justify-content: space-evenly;
}

.todaybtn {
  padding-right: 10px;
  background: transparent;
  width: 100px;
  padding: 10px;
  font-size: 16px;
  opacity: 1.8;
}
