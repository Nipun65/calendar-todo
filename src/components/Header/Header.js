import { useEffect, useState } from "react";
import styles from "./Header.module.css";

const Header = (props) => {
  let [year, setYear] = useState(props.date.year);
  let [month, setMonth] = useState(props.date.month);
  useEffect(
    (value) => {
      setMonth(props.date.month);
      setYear(props.date.year);
    },
    [props.date.month, props.date.year]
  );
  const Month = [
    "January",
    "February",
    "March",
    "April",
    "May",
    "June",
    "July",
    "August",
    "September",
    "October",
    "November",
    "December",
  ];
  let dateObj = {
    year: year,
    month: month,
    date: props.date.date,
  };

  const backYearHandler = () => {
    if (
      props.view === "months" ||
      props.view === "years" ||
      props.view === "multiyears"
    ) {
      props.setValue("back");
    }
    setYear(--year);
    dateObj.year = year;
    props.setValue({ ...dateObj });
  };
  const forwardYearHandler = () => {
    if (
      props.view === "months" ||
      props.view === "years" ||
      props.view === "multiyears"
    ) {
      props.setValue("forward");
    }
    setYear(++year);
    dateObj.year = year;
    props.setValue({ ...dateObj });
  };

  const backMonthHandler = () => {
    if (
      props.view === "months" ||
      props.view === "years" ||
      props.view === "multiyears"
    ) {
      props.setValue("backmonth");
    } else {
      if (month === 0) {
        setMonth(11);
        setYear(--year);
        dateObj.month = 11;
        dateObj.year = year;
      } else {
        setMonth(--month);
        dateObj.month = month;
      }
      props.setValue({ ...dateObj });
    }
  };
  const forwardMonthHandler = () => {
    if (
      props.view === "months" ||
      props.view === "years" ||
      props.view === "multiyears"
    ) {
      props.setValue("forwardmonth");
    } else {
      if (month === 11) {
        setMonth(0);
        setYear(++year);
        dateObj.month = 0;
        dateObj.year = year;
      } else {
        setMonth(++month);
        dateObj.month = month;
      }
      props.setValue({ ...dateObj });
    }
  };
  const viewHandler = (view) => {
    if (props.view === "date") {
      props.setView("months");
    } else if (props.view === "months") {
      props.setView("years");
    } else if (props.view === "years") {
      props.setView("multiyears");
    }
  };

  return (
    <div className={styles["header"]}>
      <div className={styles["btnHover"]}>
        {props.view !== "multiyears" && (
          <button
            onClick={backYearHandler}
            type="button"
            className={styles["btn"]}
          >
            {"<<"}
          </button>
        )}
        <button
          onClick={backMonthHandler}
          type="button"
          className={styles["btn"]}
        >
          {"<"}
        </button>
      </div>
      <div
        className={props.view !== "multiyears" ? styles["btnHover"] : ""}
        style={{ minWidth: "50px", display: "flex", alignItems: "center" }}
      >
        <button
          onClick={viewHandler}
          disabled={props.view === "multiyears"}
          className={styles["btn"]}
        >
          {props.view === "date" && <div>{Month[month]}</div>}
          {props.view === "years" &&
            (year % 10 !== 0 ? (
              <div>
                {year - (year % 10) + 1} - {year - (year % 10) + 10}
              </div>
            ) : (
              <div>
                {year - 9}-{year}
              </div>
            ))}

          {(props.view === "date" || props.view === "months") && (
            <div>{year}</div>
          )}
          {props.view === "multiyears" && (
            <div>
              {year - (year % 100) + 1} - {year - (year % 100) + 100}
            </div>
          )}
        </button>
      </div>
      <div className={styles["btnHover"]}>
        <button
          onClick={forwardMonthHandler}
          type="button"
          className={styles["btn"]}
        >
          {">"}
        </button>
        {props.view !== "multiyears" && (
          <button
            onClick={forwardYearHandler}
            type="button"
            className={styles["btn"]}
          >
            {">>"}
          </button>
        )}
      </div>
    </div>
  );
};
export default Header;
