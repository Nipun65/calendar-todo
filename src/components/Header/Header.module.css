.header {
  background-color: aqua;
  display: flex;
  justify-content: space-between;
  align-items: center;
  height: 70px;
}

.btn {
  background-color: transparent;
  border: none;
  cursor: pointer;
  font-size: 23px;
}

.btnHover :hover {
  background-color: #72a3a6;
}
