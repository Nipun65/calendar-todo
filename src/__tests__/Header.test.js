import renderer from 'react-test-renderer';
import Header from '../components/Common/Header/Header';

it('should match the snapshot', () => {
  const tree = renderer.create(<Header />).toJSON();
  expect(tree).toMatchSnapshot();
});
