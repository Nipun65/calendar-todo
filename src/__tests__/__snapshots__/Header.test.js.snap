// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot 1`] = `
<div
  className="header"
>
  <div
    className="btnHover first-div"
  >
    <button
      className="btn"
      onClick={[Function]}
      type="button"
    >
      &lt;&lt;
    </button>
    <button
      className="btn"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      &lt;
    </button>
    <div
      className="todaybtn-div"
    >
      <button
        className="todaybtn"
        onClick={[Function]}
        type="button"
      >
        Today
      </button>
    </div>
  </div>
  <div
    className="
         btnHover second-div"
  >
    <button
      className="btn"
      disabled={false}
      onClick={[Function]}
      type="button"
    />
  </div>
  <div
    className="third-div btnHover"
  >
    <div
      className="dropdown"
    >
      <div
        className="dropdown"
      >
        <button
          className="titlebutton"
          onClick={[Function]}
          type="button"
        >
          Month
        </button>
      </div>
    </div>
    <button
      className="btn"
      onClick={[Function]}
      type="button"
    >
      &gt;
    </button>
    <button
      className="btn"
      onClick={[Function]}
      type="button"
    >
      &gt;&gt;
    </button>
  </div>
</div>
`;
